RecoBTag/DeepFlavour/plugins/DeepFlavourTagInfoProducer.cc

```diff
@@ -0,0 +1,361 @@
+          }
+        }
+    }
+
+    // sort collections (open the black-box if you please) 
+    std::sort(c_sorted.begin(),c_sorted.end(),
+      btagbtvdeep::SortingClass<std::size_t>::compareByABCInv);
+    std::sort(n_sorted.begin(),n_sorted.end(),
+      btagbtvdeep::SortingClass<std::size_t>::compareByABCInv);
+
+    std::vector<size_t> c_sortedindices,n_sortedindices;
+   
+    // this puts 0 everywhere and the right position in ind 
+    c_sortedindices=btagbtvdeep::invertSortingVector(c_sorted);
+    n_sortedindices=btagbtvdeep::invertSortingVector(n_sorted);
+
+    // set right size to vectors
+    features.c_pf_features.clear();
+    features.c_pf_features.resize(c_sorted.size());
+    features.n_pf_features.clear();
+    features.n_pf_features.resize(n_sorted.size());
+
+
+
+  for (unsigned int i = 0; i <  jet.numberOfDaughters(); i++){
+
+    // get pointer and check that is correct
+    auto cand = dynamic_cast<const reco::Candidate *>(jet.daughter(i));
+    if(!cand) continue;
+    // candidates under 950MeV are not considered
+    // might change if we use also white-listing
+    if (cand->pt()<0.95) continue;
+
+    auto packed_cand = dynamic_cast<const pat::PackedCandidate *>(cand);
+    auto reco_cand = dynamic_cast<const reco::PFCandidate *>(cand);
+
+    // need some edm::Ptr or edm::Ref if reco candidates
+    reco::PFCandidatePtr reco_ptr;
+    if (pf_jet) {
+      reco_ptr = pf_jet->getPFConstituent(i);
+    } else if (pat_jet && reco_cand) {
+      reco_ptr = pat_jet->getPFConstituent(i);
+    }
+    // get PUPPI weight from value map
+    float puppiw = 1.0; // fallback value
+    if (reco_cand && use_puppi_value_map_) {
+      puppiw = (*puppi_value_map)[reco_ptr];
+    } else if (reco_cand) {
+      edm::LogWarning("MissingFeatures") << "PUPPI value map missing. "
+                                         << puppiw << " will be used as default";
+    } 
+
+    float drminpfcandsv = btagbtvdeep::mindrsvpfcand(svs_unsorted, cand);
+    
+    if (cand->charge() != 0) {
+      // is charged candidate
+      auto entry = c_sortedindices.at(i);
+      // get cached track info
+      auto & trackinfo = trackinfos.at(i);
+      // get_ref to vector element
+      auto & c_pf_features = features.c_pf_features.at(entry);
+      // fill feature structure 
+      if (packed_cand) {
+        btagbtvdeep::ChargedCandidateConverter::PackedCandidateToFeatures(packed_cand, jet, trackinfo, 
+                                                                          drminpfcandsv, c_pf_features);
+      } else if (reco_cand) {
+        // get vertex association quality
+        int pv_ass_quality = 0; // fallback value
+        if (use_pvasq_value_map_) {
+          pv_ass_quality = (*pvasq_value_map)[reco_ptr];
+        } else {
+          edm::LogWarning("MissingFeatures") << "vertex association quality map missing. "
+                                             << pv_ass_quality << " will be used as default";
```

Comment:
similar here and below: this appears to be something that will happen every reco_cand if the configuration is "wrong". This better be an exception with some strictness flag check, and otherwise the warning should be printed only once.